internal/api/handler.go

```diff
@@ -14,24 +15,34 @@ import (
 type contextHandlerFunc func(c *Context, w http.ResponseWriter, r *http.Request)
 
 type contextHandler struct {
-	context *Context
-	handler contextHandlerFunc
+	context     *Context
+	handler     contextHandlerFunc
+	handlerName string
 }
 
 func (h contextHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	start := time.Now()
+	ww := NewWrappedWriter(w)
 	context := h.context.Clone()
 	context.RequestID = model.NewID()
 	context.Logger = context.Logger.WithFields(log.Fields{
+		"handler": h.handlerName,
+		"method":  r.Method,
 		"path":    r.URL.Path,
 		"request": context.RequestID,
 	})
 
-	h.handler(context, w, r)
+	h.handler(context, ww, r)
+
+	elapsed := float64(time.Since(start)) / float64(time.Second)
+	context.Metrics.ObserveAPIEndpointDuration(h.handlerName, r.Method, ww.StatusCode(), elapsed)
```

Comment:
Wouldn't it make more sense to use `r.URL.Path` instead of `handlerName`? I think this might be error prone. It is quite possible that we copy and paste the handler and only change the actual function not the name. Also when changing function name we might forget to update the handler name, and in case we update the handler name we suddenly created new series even tho the API path will remain the same.

Reply:
So I agree that it would probably be nicer to be able to use `r.URL.Path` here, but I ran into the issue I noted above.

In order to standardize prometheus labels for the different endpoints,
the api handler name is registered and used. This is done instead
of the request URL path values because many endpoints contain
IDs which end up making them unique for each resource.

When using the Path value, any path that contained a resource ID ended up being a unique label. I attempted to keep the path value and use regex to standardize the unique IDs, but it was a bit messy. I changed tactics and used the handler name instead, which is what Mattermost server does. I agree that it isn't quite as nice, but believe it's probably the best solution for us. What do you think?

Reply:
Did a quick update to derive the func name via reflect. This removes the handler name duplication issue and maintains common metric labels.